File: leetcode/rust-leetcode/src/322.coin-change.rs
impl Solution {
    pub fn coin_change(mut coins: Vec<i32>, amount: i32) -> i32 {
        if amount == 0 {return 0;}
        let amount = amount as usize;
        let mut counts = vec![i32::MAX; amount + 1];
        counts[0] = 0;
        let coins = coins.iter().map(|&c| c as usize).collect::<Vec<_>>();
        for a in 1..counts.len() {
            for &coin in coins.iter() {
                if coin > a {
                    continue;
                }
                if counts[a-coin] != i32::MAX {
                    counts[a] = counts[a].min(counts[a-coin] + 1);
                }
            }
        }
        match counts[amount] {
            i32::MAX => -1,
            c @ _ => c
        }
    }
}

File: leetcode/python/322.coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        coins.sort()
        dp[0] = 0
        for a in range(1, amount + 1):
            for coin in coins:
                if coin > a:
                    break
                dp[a] = min(dp[a], dp[a - coin] + 1)
        return -1 if dp[amount] == float('inf') else dp[amount]

